### ISSUE TITLE
Deployment of sub-charts from requirements.yaml

### ISSUE THREAD

**Issue by hobti01** [kubernetes/helm#3067]
We're seeing some unexpected behaviour from Helm when deploying Charts with dependencies and sub-dependencies: could the expected behaviour please be documented?

Given the following Chart structure (Chart A `requirements.yaml` contains B, C, D, Chart B contains C with an alias)
```
A |-- B |-- C (1)
  |     |-- C (2)
  |
  |-- C (3)
  |-- D
```

We observe:
- All instances/aliases of Chart `C` are deployed - `C (1)`, `C (2)`, and `C (3)` are deployed in parallel.
- Chart `B` is deployed
- Chart `D` is deployed

Unexpected behaviour:
1. All instances of a Chart are deployed in parallel, regardless of position in the hierarchy.
2. Sub-charts are *not* deployed in parallel.
3. Sub-charts are deployed in the order listed in `requirements.yaml` (except for the parallelism mentioned in 1)

Could you confirm that this behaviour is `as designed`?

We'd really like to see parallel deployment of all components as a default - with ordering of deployment or dependencies only when explicitly configured.  Is there any work in this area?

**Comment by bacongobbler** [comment]
Yes, this behaviour is as designed. We do this because certain resource types in kubernetes rely on others being set up, e.g. if you deploy a deployment before a service is up, the environment variable SERVICENAME_SERVICE_HOST will not be present in the deployment's pods.

**Comment by hobti01** [comment]
I don't follow.  This is not an issue of resource types, but an issue of Chart hierarchy and ordering.

If sub-charts are deployed before parents - then `C1` and `C2` should be deployed before `B`, but `C3` should not.
If the order in `requirements.yaml` matters, then `B` should be deployed before `C3`.

Unfortunately, the ordering is unclear - can you clarify or point to the documentation describing the expected behaviour?

**Comment by hobti01** [comment]
We are seeing the same issue in reverse when undeploying "layered" charts: the sub-chart (i.e. backend database) is undeployed *before* the front-end.  This puts the front-end into a CrashLoop, helm cannot undeploy the crashing statefulset, node CPU spikes to 100, the node becomes unavailable, and much hilarity ensues.

There are many potential issues in that scenario, but the root cause is deleting the sub-chart before the parent chart.

@bacongobbler I agree that the resource types must be ordered within a single Chart, but I don't agree that this is appropriate with sub-charts.

**Comment by skyfirst93** [comment]
Any Update
